buildSrc/src/main/java/com/gradle/publish/DownloadMavenArtifactsAndPublishToGradlePluginPortal.java

```diff
@@ -0,0 +1,311 @@
+package com.gradle.publish;
+
+import com.gradle.publish.protocols.v1.models.publish.PublishArtifact;
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.GradleException;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.TaskAction;
+
+/**
+ * This task is used to download Maven artifacts and publish them to the Gradle Plugin Portal.
+ */
+public abstract class DownloadMavenArtifactsAndPublishToGradlePluginPortal extends DefaultTask {
+
+  private static final Logger LOGGER = Logging.getLogger(DownloadMavenArtifactsAndPublishToGradlePluginPortal.class);
+
+  private static final Set<String> REQUIRED_ARTIFACT_TYPES = Set.of(
+    "pom",
+    "pom.asc",
+    "jar",
+    "jar.asc",
+    "jar:sources",
+    "jar.asc:sources",
+    "jar:javadoc",
+    "jar.asc:javadoc",
+    "module",
+    "module.asc");
+
+  @Input
+  private String pluginConfigurationName;
+
+  @Input
+  private String mavenSourceRepositoryUrl;
+
+  @Input
+  private String mavenAuthorizationToken;
+
+  @Input
+  private String groupId;
+
+  @Input
+  private String artifactId;
+
+  @Input
+  private String version;
+
+  @Input
+  private boolean validationOnly;
+
+  @Input
+  private boolean simulatePublication;
+
+  /**
+   * Main method to execute the task.
+   * @throws Exception if any error occurs during task execution
+   */
+  @TaskAction
+  public void executeTask() throws InterruptedException {
+    downloadRequiredArtifactsInBuildFolder();
+    validatePublishContent();
+    if (simulatePublication) {
+      try (var localGradlePortal = new SimpleGradlePluginPortalServer()) {
+        overrideSystemProperty("gradle.portal.url", localGradlePortal.getServerUrl(), this::executePublishTask);
+      }
+    } else {
+      executePublishTask();
+    }
+  }
+
+  public static void overrideSystemProperty(String propertyName, String newValue, Runnable runnable) {
```

Comment:
This method does not seem to be used outside of the task. Can we make this method be private?
```suggestion
  private static void overrideSystemProperty(String propertyName, String newValue, Runnable runnable) {
```